File: CSharp/ByTier/D/2151_D_Grid_Counting.cs
// Problem: CF 2151 D - Grid Counting
// https://codeforces.com/contest/2151/problem/D

/* 
 * Problem: D. Grid Counting
 * 
 * Purpose:
 *   Count the number of ways to color cells in an n x n grid such that:
 *   1. Each row k has exactly a[k] black cells.
 *   2. There is exactly one black cell in each row with max(x_i, y_i) = k.
 *   3. There is exactly one black cell in each row with max(x_i, n+1-y_i) = k.
 * 
 * Approach:
 *   - This problem involves combinatorics and dynamic programming on a tree-like structure.
 *   - The constraints define that for each row k:
 *     - There are exactly a[k] black cells.
 *     - There is one special cell in this row such that max(x_i, y_i) = k.
 *     - There is also exactly one special cell such that max(x_i, n+1-y_i) = k.
 *   - These two special rows form a structure akin to a Catalan number variant, where
 *     we can model the arrangement of valid pairs using combinations.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */

﻿using System.Text;
using System.IO;
using System;
class Program {
  static long Solve() {
    int n = Next();
    int[] a = new int[n];
    long s = 0;
    for (int i = 0; i < n; i++)
      s += a[i] = Next(); // Read the array and compute sum
    if (s != n)
      return 0; // If total black cells is not equal to n, invalid configuration
    long ans = 1;
    for (int i = (n + 1) / 2; i < n; i++)
      if (a[i] > 0)
        return 0; // Check that rows from middle onwards have no black cells
    for (int i = (n + 1) / 2 - 1, c = 1 + (n + 1) % 2; i >= 0; i--, c += 2) {
      if (a[i] > c)
        return 0; // Check if number of black cells exceeds allowed maximum
      ans = ans * GetCFact(c, a[i]) % mod; // Compute combination C(c, a[i])
      c -= a[i]; // Decrement the available slots
    }
    return ans;
  }
  private const int mod = 998244353;
  private static long[] fact;
  private static long[] fact1;
  private static void InitFact(int n) {
    fact = new long[n];
    fact1 = new long[n];
    fact[0] = 1;
    for (int i = 1; i < fact.Length; i++) {
      fact[i] = (fact[i - 1] * i) % mod; // Precompute factorials
    }
    fact1[n - 1] = Pow(fact[n - 1], mod - 2);
    for (int i = fact1.Length - 2; i >= 0; i--) {
      fact1[i] = (fact1[i + 1] * (i + 1)) % mod; // Precompute inverse factorials
    }
  }
  private static long GetCFact(int n, int k) {
    return (((fact[n] * fact1[k]) % mod) * fact1[n - k]) % mod; // Combination C(n, k)
  }
  private static long Pow(long a, int k) {
    long r = 1;
    while (k > 0) {
      if ((k & 1) == 1) {
        r = (r * a) % mod;
      }
      a = (a * a) % mod;
      k >>= 1;
    }
    return r;
  }
  static void Main(string[] args) {
    InitFact(200001); // Initialize factorial arrays up to 200001
    long t = Next();
    for (int i = 0; i < t; i++)
      writer.WriteLine(Solve());
    writer.Flush();
  }
  static int Next() {
    int c;
    int m = 1;
    int res = 0;
    do {
      c = reader.Read();
      if (c == '-')
        m = -1;
    } while (c < '0' || c > '9');
    res = c - '0';
    while (true) {
      c = reader.Read();
      if (c < '0' || c > '9')
        return m * res;
      res *= 10;
      res += c - '0';
    }
  }
  static readonly StreamReader reader = new StreamReader(
      Console.OpenStandardInput(1024 * 10), Encoding.ASCII, false, 1024 * 10);
  static readonly StreamWriter writer = new StreamWriter(
      Console.OpenStandardOutput(1024 * 10), Encoding.ASCII, 1024 * 10);
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByTier/D/2151_D_Grid_Counting.py
# Problem: CF 2151 D - Grid Counting
# https://codeforces.com/contest/2151/problem/D

"""
D. Grid Counting
Algorithms/Techniques: Dynamic Programming with Combinatorics, Preprocessing Factorials and Modular Inverses

Time Complexity: O(n) per test case (preprocessing factorials + constant time per query)
Space Complexity: O(n) for storing factorials and inverse factorials

The problem asks us to count the number of valid ways to color cells in an n×n grid such that:
- Each row k has exactly a[k] black cells.
- There exists exactly one cell (xi, yi) in each row where max(xi, yi) = k.
- There exists exactly one cell (xi, yi) in each row where max(xi, n+1-yi) = k.

This is solved using dynamic programming and combinatorics to count valid placements,
where we simulate the process of placing black cells row by row, ensuring constraints
are satisfied at each step. The use of precomputed factorials and modular inverses is for
efficient binomial coefficient computation.

"""

nt = int(input())
M = 998_244_353
f = [1] * 200_033
for i in range(1, 200_033):
    f[i] = (i * f[i - 1]) % M
fi = [pow(x, -1, M) for x in f]  # Precompute modular inverses of factorials

for _ in range(nt):
    n = int(input())
    xs = [int(i) for i in input().split()]
    
    if xs[0] < 2:
        print(0)
        continue
    if sum(xs) != n:
        print(0)
        continue
    if sum(xs[(n - 1) // 2 + 1 :]) != 0:
        print(0)
        continue

    ans = 1
    r = 2 if n % 2 == 0 else 1  # Starting "available slots" for placement
    for i in range(((n - 1) // 2), -1, -1):
        if xs[i] > r:
            # Impossible to place more black cells than available positions
            ans *= 0
        else:
            # Compute binomial coefficient C(r, xs[i]) mod M and update answer
            ans *= (f[r] * fi[xs[i]] * fi[r - xs[i]]) % M
            ans = ans % M
        r += 2 - xs[i]  # Update number of available slots for next iteration
    
    print(ans)


# https://github.com/VaHiX/CodeForces/